### PR TITLE
fix configure_spm.py

### PR DESCRIPTION
**Pull request by jeromedockes**

### CODE REVIEW

File: pypreprocess/configure_spm.py [hunk 1]
```diff
@@ -6,186 +6,193 @@
...
+def _check_nipype_version():
     if distutils.version.LooseVersion(
             nipype.__version__).version < [0, 9]:
-        warnings.warn(("Nipype version %s too old. No support for "
-                       "precompiled SPM!") % nipype.__version__)
-        return
+        return False
+    return True
+
+def _find_or_warn(loc, check, msg=None, warner=_logger.warn, recursive=False):
```
Comment (dohmatob) [inline pypreprocess/configure_spm.py]: Missing doc (for maintainance by other dev, etc.)

File: pypreprocess/configure_spm.py [hunk 2]
```diff
@@ -6,186 +6,193 @@
...
+        for (dirpath, dirnames, filenames) in os.walk(loc):
+            for name in itertools.chain(dirnames, filenames):
+                found_path = os.path.join(dirpath, name)
+                if check(found_path):
+                    return found_path
+    if msg is not None:
+        warner(msg)
+    return None
+
+def _find_dep_loc(specified_loc=None, exported_name=None, default_locs=None,
```
Comment (dohmatob) [inline pypreprocess/configure_spm.py]: Missing doc (for maintenance by other dev, etc.)

File: pypreprocess/configure_spm.py [hunk 3]
```diff
@@ -6,186 +6,193 @@
...
+                              recursive=recursive, warner=_logger.debug)
+        if found is not None:
+            return found
+
+    return None
+
+def _is_executable(file_name):
+    return os.access(file_name, os.X_OK)
+
+def _find_spm_mcr(spm_mcr):
```
Comment (dohmatob) [inline pypreprocess/configure_spm.py]: missing doc

File: pypreprocess/configure_spm.py [hunk 4]
```diff
@@ -6,186 +6,193 @@
...
+    return(os.path.isdir(dir_path) and 'tpm' in os.listdir(dir_path))
+
+def _find_spm_dir(spm_dir):
+    return _find_dep_loc(
+        spm_dir, 'SPM_DIR', DEFAULT_SPM_DIRS,
+        check=_is_spm_dir,
+        msg_prefix='SPM directory could not be found in',
+        recursive=True)
+
+def _find_spm_mcr_and_spm_dir(spm_mcr, spm_dir):
```
Comment (dohmatob) [inline pypreprocess/configure_spm.py]: doc

File: pypreprocess/configure_spm.py [hunk 5]
```diff
@@ -6,186 +6,193 @@
...
+    if spm_mcr is None:
+        return None
+
+    spm_dir = _find_spm_dir(spm_dir)
+    if spm_dir is not None:
+        _logger.info('found SPM directory to be "{}"'.format(spm_dir))
+        return spm_dir
+    return None
+
+def _find_matlab_exec_and_spm_dir(spm_dir, matlab_exec):
```
Comment (dohmatob) [inline pypreprocess/configure_spm.py]: doc

File: pypreprocess/configure_spm.py [hunk 6]
```diff
@@ -6,186 +6,193 @@
...
+        return False
+    return True
+
+def _find_or_warn(loc, check, msg=None, warner=_logger.warn, recursive=False):
+    if loc is None:
+        return None
+    loc = os.path.abspath(os.path.expanduser(loc))
+    if check(loc):
+        return loc
+    if(recursive and os.path.isdir(loc)):
```
Comment (dohmatob) [inline pypreprocess/configure_spm.py]: We need to test the code block that follows. Many things can go wrong here.

**Comment by dohmatob** [comment]
OK, thanks for this PR. I'm looking at it. Small remark: it's a good idea to work directly on your master branch :) I'd create a branch called **fix-spm-config**, e.g.

**Comment by dohmatob** [comment]
The log-file writen during the SPM configuration is weird. For example, it contains repeated messages.

```
2016-06-03 11:10:39,015:INFO: setting SPM MCR backend: set matlab executable path to "/home/elvis/opt/spm8/spm8.sh" and "use_mcr" to True
2016-06-03 11:10:39,080:INFO: found SPM directory to be "/home/elvis/opt/spm8/spm8/spm8_mcr/spm8"
2016-06-03 11:10:39,081:INFO: setting SPM MCR backend: set matlab executable path to "/home/elvis/opt/spm8/spm8.sh" and "use_mcr" to True
2016-06-03 11:10:39,156:INFO: found SPM directory to be "/home/elvis/opt/spm8/spm8/spm8_mcr/spm8"
2016-06-03 11:10:39,541:INFO: setting SPM MCR backend: set matlab executable path to "/home/elvis/opt/spm8/spm8.sh" and "use_mcr" to True
2016-06-03 11:10:39,605:INFO: found SPM directory to be "/home/elvis/opt/spm8/spm8/spm8_mcr/spm8"
2016-06-03 11:10:39,723:INFO: setting SPM MCR backend: set matlab executable path to "/home/elvis/opt/spm8/spm8.sh" and "use_mcr" to True
2016-06-03 11:10:39,787:INFO: found SPM directory to be "/home/elvis/opt/spm8/spm8/spm8_mcr/spm8"
2016-06-03 11:10:40,795:INFO: setting SPM MCR backend: set matlab executable path to "/home/elvis/opt/spm8/spm8.sh" and "use_mcr" to True
2016-06-03 11:10:40,865:INFO: found SPM directory to be "/home/elvis/opt/spm8/spm8/spm8_mcr/spm8"
2016-06-03 11:10:40,993:INFO: setting SPM MCR backend: set matlab executable path to "/home/elvis/opt/spm8/spm8.sh" and "use_mcr" to True

```

**Comment by dohmatob** [comment]
OK, it seems there is no fire. There are redundant messages because the spm config is being done in each node of the pipeline.

**Comment by jeromedockes** [comment]
yes, just checked that (but if it's not useful we can remove logging anyway):

2016-06-03 13:48:01,202:INFO: _configure_spm call # 0
2016-06-03 13:48:01,203:INFO: setting SPM MCR backend: set matlab executable path to "/home/student/opt/spm8/spm8.sh" and "use_mcr" to True
2016-06-03 13:48:01,213:INFO: found SPM directory to be "/home/student/opt/spm8/spm8/spm8_mcr/spm8"
2016-06-03 13:48:01,214:INFO: _configure_spm call # 1
2016-06-03 13:48:01,215:INFO: setting SPM MCR backend: set matlab executable path to "/home/student/opt/spm8/spm8.sh" and "use_mcr" to True
2016-06-03 13:48:01,216:INFO: found SPM directory to be "/home/student/opt/spm8/spm8/spm8_mcr/spm8"
2016-06-03 13:48:01,593:INFO: _configure_spm call # 2
2016-06-03 13:48:01,594:INFO: setting SPM MCR backend: set matlab executable path to "/home/student/opt/spm8/spm8.sh" and "use_mcr" to True
2016-06-03 13:48:01,594:INFO: found SPM directory to be "/home/student/opt/spm8/spm8/spm8_mcr/spm8"
2016-06-03 13:48:01,797:INFO: _configure_spm call # 3
2016-06-03 13:48:01,797:INFO: setting SPM MCR backend: set matlab executable path to "/home/student/opt/spm8/spm8.sh" and "use_mcr" to True
2016-06-03 13:48:01,798:INFO: found SPM directory to be "/home/student/opt/spm8/spm8/spm8_mcr/spm8"
2016-06-03 13:48:02,777:INFO: _configure_spm call # 4
2016-06-03 13:48:02,777:INFO: setting SPM MCR backend: set matlab executable path to "/home/student/opt/spm8/spm8.sh" and "use_mcr" to True
2016-06-03 13:48:02,778:INFO: found SPM directory to be "/home/student/opt/spm8/spm8/spm8_mcr/spm8"
2016-06-03 13:48:03,016:INFO: _configure_spm call # 5
2016-06-03 13:48:03,017:INFO: setting SPM MCR backend: set matlab executable path to "/home/student/opt/spm8/spm8.sh" and "use_mcr" to True
2016-06-03 13:48:03,018:INFO: found SPM directory to be "/home/student/opt/spm8/spm8/spm8_mcr/spm8"

**Comment by dohmatob** [comment]
LGTM. I'll do a final pass and merge tomorrow. Thanks.

**Comment by dohmatob** [comment]
OK, merging. A few fixups will probably follow, in the form of much smaller PRs. Thanks @jeromedockes for the great job!

**Pull request merged**